File: leetcode/C++/array-of-doubled-pairs.cpp
class Solution {
public:
    bool canReorderDoubled(vector<int>& arr) {
        int t=0;
        unordered_map<int,int> um;
        sort(arr.begin(),arr.end());
        for(int &n : arr) um[n]+=1;
        for(int &n : arr){
            if(--um[n]<0) continue;
            else if(n<0 && n%2) return false;
            t = (n>=0)? 2*n : n/2; 
            if(--um[t]<0) return false;
        }
        return true;
    }
};

File: leetcode/Python/array-of-doubled-pairs.py
class Solution:
    def canReorderDoubled(self, arr: List[int]) -> bool:
        um, t = defaultdict(int), 0
        arr.sort()
        for n in arr:
            um[n]+=1
        for n in arr:
            if um[n]==0:
                continue
            elif n<0 and n%2>0:
                return False
            t = 2*n if n>=0 else n//2
            if (n==t and um[n]<2) or (n!=t and um[t]==0):
                return False
            um[n]-=1
            um[t]-=1
        return True